plugin/user.c

```diff
@@ -100,26 +107,214 @@ static int load_config_user(void) {
+    char b = (version >> 20) & 0xF;
+    char c = (version >> 16) & 0xF;
+    char d = (version >> 12) & 0xF;
+    string[0] = '0' + a;
+    string[1] = '.';
+    string[2] = '0' + b;
+    string[3] = '0' + c;
+    string[4] = '\0';
+    if (d) {
+      string[4] = '0' + d;
+      string[5] = '\0';
+    }
+    return 1;
+  }
+  return 0;
+}
+
+static tai_hook_ref_t g_sceRegMgrGetKeyStr_SceSystemSettingsCore_hook;
+static int sceRegMgrGetKeyStr_SceSystemSettingsCore_patched(const char *category, const char *name, char *string, int length) {
+  if (sceClibStrncmp(category, "/CONFIG/HENKAKU", 15) == 0) {
+    if (sceClibStrncmp(name, "spoofed_version", 15) == 0) {
+      if (string != NULL) {
+        load_config_user();
+        sceClibMemset(string, 0, length);
+        if (!build_version_string(config.spoofed_version, string, length))
+          build_version_string(SPOOF_VERSION, string, length);
+      }
+    }
+    return 0;
+  }
+  return TAI_CONTINUE(int, g_sceRegMgrGetKeyStr_SceSystemSettingsCore_hook, category, name, string, length);
+}
+
+#define IS_DIGIT(i) (i >= '0' && i <= '9')
+static tai_hook_ref_t g_sceRegMgrSetKeyStr_SceSystemSettingsCore_hook;
+static int sceRegMgrSetKeyStr_SceSystemSettingsCore_patched(const char *category, const char *name, const char *string, int length) {
+  if (sceClibStrncmp(category, "/CONFIG/HENKAKU", 15) == 0) {
+    if (sceClibStrncmp(name, "spoofed_version", 15) == 0) {
+      if (string != NULL) {
+        if (IS_DIGIT(string[0]) && string[1] == '.' && IS_DIGIT(string[2]) && IS_DIGIT(string[3])) {
+          char a = string[0] - '0';
+          char b = string[2] - '0';
+          char c = string[3] - '0';
+          char d = IS_DIGIT(string[4]) ? string[4] - '0' : '\0';
+          config.spoofed_version = ((a << 24) | (b << 20) | (c << 16) | (d << 12));
+          save_config_user();
+          henkaku_reload_config();
+        }
+      }
+    }
+    return 0;
+  }
+  return TAI_CONTINUE(int, g_sceRegMgrSetKeyStr_SceSystemSettingsCore_hook, category, name, string, length);
+}
+
+typedef struct {
+  int size;
+  const char *name;
+  int type;
+  int unk;
+} SceRegMgrKeysInfo;
+
+static tai_hook_ref_t g_sceRegMgrGetKeysInfo_SceSystemSettingsCore_hook;
+static int sceRegMgrGetKeysInfo_SceSystemSettingsCore_patched(const char *category, SceRegMgrKeysInfo *info, int unk) {
+  if (sceClibStrncmp(category, "/CONFIG/HENKAKU", 15) == 0) {
+    if (info) {
+      if (sceClibStrncmp(info->name, "spoofed_version", 15) == 0) {
+        info->type = 0x00030001;
+      } else {
+        info->type = 0x00040000;
+      }
+    }
+
+    return 0;
+  }
+  return TAI_CONTINUE(int, g_sceRegMgrGetKeysInfo_SceSystemSettingsCore_hook, category, info, unk);
+}
+
+static tai_hook_ref_t g_ScePafMisc_19FE55A8_SceSettings_hook;
+static int ScePafMisc_19FE55A8_SceSettings_patched(int a1, void *xml_buf, int xml_size, int a4) {
+  if (sceClibStrncmp(xml_buf+82, "system_settings_plugin", 22) == 0 && (82+22) < xml_size) {
```

Comment:
The order of the two checks should be reversed since we want (82+22) < xml_size to short circuit and not run strncmp.